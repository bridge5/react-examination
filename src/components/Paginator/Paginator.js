import React, { useState, useEffect } from 'react';
import styles from './Paginator.scss';
import { Range, Set } from 'immutable';
import classnames from 'classnames';

export function buildPageButtons(currentPage, totalPage) {
  let range;
  if (totalPage <= 5 || currentPage <= 3) {
    range = Range(1, Set([totalPage, 5]).min() + 1);
  } else if (currentPage > 3 && totalPage - currentPage > 2) {
    range = Range(currentPage - 2, currentPage + 3);
  } else {
    range = Range(totalPage - 4, totalPage + 1);
  }
  const buttons = range
    .map(pageIndex => ({ label: String(pageIndex), value: pageIndex }))
    .toList()
    .asMutable();
  if (currentPage === 1) {
    buttons
      .rest()
      .toList()
      .unshift({ label: '1', value: 1 });
  } else if (currentPage === totalPage) {
    buttons
      .butLast()
      .toList()
      .push({ label: totalPage.toString(), value: totalPage });
  }
  return buttons.asImmutable();
}

export const Paginator = ({ cb, totalPage }) => {
  const [currentPage, setCurrentPage] = useState(1);
  const [pageButton, setPageButton] = useState(buildPageButtons(1, 20));
  console.log(pageButton)

  const onClickPage = pageNumber => {
    setCurrentPage(pageNumber);
    cb && cb();
  };

  useEffect(() => {
    if(currentPage > pageButton.get(4).value || currentPage < pageButton.get(0).value) {
      setPageButton(buildPageButtons(currentPage, 20))
    }

  }, [currentPage]);


  return (
    <div className={styles.wrap}>
      {currentPage > 1 && (
        <li
          onClick={() => onClickPage(currentPage - 1)}
          className={`${styles.btn}`}
        >
          Prev
        </li>
      )}
      {pageButton.map((v, i) => (
        <li
          onClick={() => onClickPage(v.value)}
          key={v.value}
          className={`${styles.btn} ${classnames({
            active: currentPage === v.value,
          })}`}
        >
          {v.label}
        </li>
      ))}
      {currentPage < 20 && (
        <li
          onClick={() => onClickPage(currentPage + 1)}
          className={`${styles.btn}`}
        >
          Prev
        </li>
      )}
    </div>
  );
};
