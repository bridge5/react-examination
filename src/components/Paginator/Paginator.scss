:local(.wrap) {
    width: 100%;
    text-align: center;
    background: #fff;
    padding: 10px 0;
}

:local(.btn) {
  position: relative;
  display: inline-block;
  padding: 6px 12px;
  margin: 0 5px;
  line-height: 1.428571429;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  background-image: none;
  border: 1px solid transparent;
  border-radius: 4px;
  background-color: #d9534f;
  color: white;
  outline: 0;
  text-decoration: none;
  transition: all .1s linear;
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;

  &:hover {
    background-color: darken(#d9534f, 5%);
  }
  &.active {
    background-color: #fff;
    color: #222222;
    border: 1px solid #222222;
  }
}
