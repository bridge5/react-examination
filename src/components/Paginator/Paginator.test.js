import React from 'react';
import { Paginator } from "./Paginator";
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
   const wrapper = renderer.create(<Paginator totalPage={20} currentPage={1} />).toJSON();
   expect(wrapper).toMatchSnapshot();
});
