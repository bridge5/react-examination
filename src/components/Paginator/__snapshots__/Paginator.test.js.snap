// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div>
  <li
    className="undefined active"
    onClick={[Function]}
  >
    1
  </li>
  <li
    className="undefined "
    onClick={[Function]}
  >
    2
  </li>
  <li
    className="undefined "
    onClick={[Function]}
  >
    3
  </li>
  <li
    className="undefined "
    onClick={[Function]}
  >
    4
  </li>
  <li
    className="undefined "
    onClick={[Function]}
  >
    5
  </li>
  <li
    className="undefined"
    onClick={[Function]}
  >
    Next
  </li>
</div>
`;
